src/OpenSpaceToolkit/Astrodynamics/Trajectory/TrajectorySegment.cpp

```diff
@@ -0,0 +1,178 @@
+/// Apache License 2.0
+
+#include <OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.hpp>
+#include <OpenSpaceToolkit/Astrodynamics/Trajectory/TrajectorySegment.hpp>
+
+namespace ostk
+{
+namespace astro
+{
+namespace trajectory
+{
+
+using ostk::physics::time::Duration;
+
+using ostk::astro::trajectory::Propagator;
+
+TrajectorySegment::Solution::Solution(
+    const String& aName, const Array<Shared<Dynamics>>& aDynamicsArray, const Array<State>& aStates
+)
+    : name(aName),
+      dynamics(aDynamicsArray),
+      states(aStates)
+{
+}
+
+TrajectorySegment::TrajectorySegment(
+    const String& aName,
+    const TrajectorySegment::Type& aType,
+    const Shared<EventCondition>& anEventConditionSPtr,
+    const Array<Shared<Dynamics>>& aDynamicsArray,
+    const NumericalSolver& aNumericalSolver
+)
+    : name_(aName),
+      type_(aType),
+      eventCondition_(anEventConditionSPtr),
+      dynamics_(aDynamicsArray),
+      numericalSolver_(aNumericalSolver)
+{
+    if (eventCondition_ == nullptr)
+    {
+        throw ostk::core::error::runtime::Undefined("Event condition");
+    }
+
+    if (dynamics_.isEmpty())
+    {
+        throw ostk::core::error::runtime::Undefined("Dynamics");
+    }
+
+    if (!numericalSolver_.isDefined())
+    {
+        throw ostk::core::error::runtime::Undefined("Numerical solver");
+    }
+}
+
+std::ostream& operator<<(std::ostream& anOutputStream, const TrajectorySegment& aSegment)
+{
+    aSegment.print(anOutputStream, false);
+
+    return anOutputStream;
+}
+
+String TrajectorySegment::getName() const
+{
+    return name_;
+}
+
+Shared<EventCondition> TrajectorySegment::getEventCondition() const
+{
+    return accessEventCondition();
+}
+
+Array<Shared<Dynamics>> TrajectorySegment::getDynamics() const
+{
+    return accessDynamics();
+}
+
+NumericalSolver TrajectorySegment::getNumericalSolver() const
+{
+    return accessNumericalSolver();
+}
+
+TrajectorySegment::Type TrajectorySegment::getType() const
+{
+    return type_;
+}
+
+const Shared<EventCondition>& TrajectorySegment::accessEventCondition() const
+{
+    return eventCondition_;
+}
+
+const Array<Shared<Dynamics>>& TrajectorySegment::accessDynamics() const
+{
+    return dynamics_;
+}
+
+const NumericalSolver& TrajectorySegment::accessNumericalSolver() const
+{
+    return numericalSolver_;
+}
+
+TrajectorySegment::Solution TrajectorySegment::solve(const State& aState, const Duration& maximumPropagationDuration)
+    const
+{
+    const Propagator propagator = {
+        numericalSolver_,
+        dynamics_,
+    };
+
+    const Instant startInstant = aState.getInstant();
+
+    // TBI: Handle the case where the condition is not met
```

Comment:
Do we need to handle this in this PR or make a separate one?

Reply:
in another PR when we learn about how we want to handle this case.